File: TAP_practica1/src/main/Tests/test_mapreduce/TestMapReduce.java
package main.Tests.test_mapreduce;

import main.FaaS.Controller;
import main.MapReduce.*;

import java.io.BufferedWriter;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Map;
import java.util.List;
import java.util.concurrent.Future;
import java.util.function.Function;

public class TestMapReduce {
    public static void main(String[] args) throws FileNotFoundException {


        Controller controller = new Controller(10, 1024);
        Function<List<String>, Map<String, Integer>> func = list -> WordCount.wordCount(list);
        controller.registerAction("wordCount", func, 512);

        Function<List<String>, Integer> func2 = list -> CountWords.countWords(list);
        controller.registerAction("countWords", func2, 512);
        
        List<List<String>> textCollections = FilesReader.readFiles();
        
        Future<List<Map<String, Integer>>> wordCountResults = controller.invoke_async("wordCount", textCollections);
        Future<List<Integer>> countWordsResults = controller.invoke_async("countWords", textCollections);
        
        try {
            Map<String, Integer> finalWordCountResult = Reduce.reduceWordCount(wordCountResults.get());
            int finalCountWordsResult = Reduce.reduceCountWords(countWordsResults.get());
            System.out.println(countWordsResults.get());
            System.out.println(finalCountWordsResult);
            saveWordCountsToFile(finalWordCountResult, "src/main/Tests/test_mapreduce/results/mapReduceResults.txt");
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
        
        controller.shutdown();
    }

    private static void saveWordCountsToFile(Map<String, Integer> map, String fileName) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileName))) {
            for (Map.Entry<String, Integer> entry : map.entrySet()) {
                writer.write(entry.getKey() + ": " + entry.getValue());
                writer.newLine();
            }

        } catch (IOException e) {
            System.err.println("Error writing to file: " + e.getMessage());
        }
    }

}

File: TAP_practica2/src/main/scala/Tests/test_mapreduce/TestMapReduce.scala
package Tests.test_mapreduce

import FaaS.Controller
import FilesReader.FilesReader
import MapReduce._

import java.util.concurrent.CompletableFuture
import scala.jdk.CollectionConverters._
import scala.jdk.FunctionConverters.enrichAsJavaFunction
import java.util.{List => JavaList, Map => JavaMap}
import scala.collection.mutable
import scala.util.{Failure, Success, Try}


object TestMapReduce extends App {
  val controller = new Controller(10, 1024)

  val wordCount = new WordCount()
  val countWords = new CountWords()
  val reduce = new Reduce()

  val wordCountFunc: JavaList[String] => JavaMap[String, Integer] =
    (list: JavaList[String]) => {
      val scalaList = list.asScala.toList
      val wordCountResult = wordCount.wordCount(scalaList)
      wordCountResult.map { case (word, count) => (word, count: Integer) }.asJava
    }

  controller.registerAction("wordCount", wordCountFunc.asJava, 512)

  val countWordsFunc: JavaList[String] => Integer =
    (list: JavaList[String]) => {
      val scalaList = list.asScala.toList
      val countWordsResult = countWords.countWords(scalaList)
      countWordsResult: Integer
    }


  controller.registerAction("countWords", countWordsFunc.asJava, 512)

  val textCollections = FilesReader.readFiles()

  val wordCountResultsFuture: CompletableFuture[JavaList[JavaMap[String, Int]]] =
    controller.invoke_async("wordCount", textCollections)

  val countWordsResultsFuture: CompletableFuture[JavaList[Int]] =
    controller.invoke_async("countWords", textCollections)

  Try {
    val wordCountResults: JavaList[JavaMap[String, Int]] = wordCountResultsFuture.get()
    val countWordsResults: JavaList[Int] = countWordsResultsFuture.get()

    val scalaWordCountResults: List[mutable.Map[String, Int]] = wordCountResults.asScala.map(_.asScala).toList
    val scalaCountWordsResults: List[Int] = countWordsResults.asScala.toList



    val finalWordCountResult = reduce.reduceWordCount(scalaWordCountResults)
    val finalCountWordsResult = reduce.reduceCountWords(scalaCountWordsResults)

    val javaWordCountResults: JavaMap[String, Integer] = finalWordCountResult.map { case (k, v) => (k, v: Integer) }.asJava

    println(scalaCountWordsResults)
    println(finalCountWordsResult)
    FilesReader.saveWordCountsToFile(javaWordCountResults, "src/main/scala/Tests/test_mapreduce/results/mapReduceResults.txt")
  } match {
    case Success(_) =>
    case Failure(exception) => println(exception.getMessage)
  }

  controller.shutdown()
}
